### PR TITLE
json: yes()

### LINKED ISSUE #654: Define a public API
(Note: This issue is phrased assertively to fit into the format on a Issue Tracker. This is by no means an attempt to force anything onto the project.)

Borg should have a public, stable and documented API for other tools to consume. This can be either by defining public python APIs or a pipe-based interface with borg commands.

This would faciliate the creation of wrapper tools which work with borg, such as GUIs and anything beyond simple cronjobs.

### PR DESCRIPTION
**Pull request by enkore**
Fixes #654 

Although some minor / not overly important bits and pieces are still missing, I'd have to make a survey to find them all. Thinking recreate might not do the stats or file list right (or does it?), perhaps adding a few more tests, and working a more general intro / summary into the docs and more [comprehensive] reference material.

Also, feedback if someone wants to make a test drive would be valuable, before we freeze this interface and define compatibility policies etc.

### CODE REVIEW

**Comment by enkore** [comment]
It doesn't do any good for a frontend at all since if you have an encoding issue between the frontend and Borg (stdin) then you'll have the same problem between Borg and the frontend (stdout).

**Comment by enkore** [comment]
> If in doubt an application should probably use PYTHONIOENCODING?

Or we fix IO encoding to UTF-8 (irrespective of locale) in JSON mode, which probably makes more sense and is less error prone for downstream developers anyway.

I.e., --log-json → stderr text is always UTF-8, stdin text (so not "borg create -") is always UTF-8 (prompts, passwords), env-vars are read as UTF-8 (dito)

--json → stdout text is always UTF-8

**Comment by enkore** [comment]
merge?

File: docs/internals/frontends.rst [hunk 1]
```diff
@@ -129,10 +131,78 @@ A debug log message::
...
+*question_accepted_default*, *question_accepted_true*, *question_accepted_false* and
+*question_env_answer* types.
+
+The *message* property contains the same string displayed regularly in the same situation,
+while the *msgid* property may contain a msgid_, typically the name of the
+environment variable that can be used to override the prompt. It is the same for all JSON
+messages pertaining to the same prompt.
+
+The *is_prompt* boolean property distinguishes informational messages from prompts, it
+is true for *question_prompt* and *question_prompt_retry* types, otherwise it is false.
```
Comment (ThomasWaldmann) [inline docs/internals/frontends.rst]: hmm, is a boolean best here? it seems to be the kind of message, maybe more than 2 kinds could exist here.

Comment (ThomasWaldmann) [inline docs/internals/frontends.rst]: also you already have kind=question_prompt.

Comment (enkore) [inline docs/internals/frontends.rst]: Yeah, it's kinda redundant, mostly useful for the internal branching, maybe not so useful for external use.

**Pull request merged**